pcap4j-core/src/main/java/org/pcap4j/packet/GtpV1PduSessionContainerExtensionHeader.java

```diff
@@ -0,0 +1,490 @@
+          .append(" bytes). data: ")
+          .append(ByteArrays.toHexString(rawData, " "))
+          .append(", offset: ")
+          .append(offset)
+          .append(", length: ")
+          .append(length);
+      throw new IllegalRawDataException(sb.toString());
+    }
+
+    int lengthInRaw = (ByteArrays.getByte(rawData, LENGTH_OFFSET + offset)) & 0xFF;
+    if (lengthInRaw != GTPV1_PDU_SESSION_CONTAINER_EXTENSION_MIN_LENGTH / 4
+        && lengthInRaw != GTPV1_PDU_SESSION_CONTAINER_EXTENSION_MAX_LENGTH / 4) {
+      StringBuilder sb = new StringBuilder(100);
+      sb.append("The length filed value must be ")
+          .append(GTPV1_PDU_SESSION_CONTAINER_EXTENSION_MIN_LENGTH / 4)
+          .append(" or ")
+          .append(GTPV1_PDU_SESSION_CONTAINER_EXTENSION_MAX_LENGTH / 4)
+          .append(", but it is ")
+          .append(lengthInRaw);
+      throw new IllegalRawDataException(sb.toString());
+    }
+
+    this.length = (byte) lengthInRaw;
+
+    byte pduTypeAndSpare = ByteArrays.getByte(rawData, PDU_TYPE_AND_SPARE_OFFSET + offset);
+    this.pduType = (byte) ((pduTypeAndSpare & 0xF0) >>> 4);
+
+    if (pduType == 0) {
+      byte pppAndRqiAndQfi = ByteArrays.getByte(rawData, PPP_AND_RQI_AND_QFI_OFFSET + offset);
+      this.ppp = ((pppAndRqiAndQfi & 0x80) >>> 7) != 0;
+      this.rqi = ((pppAndRqiAndQfi & 0x40) >> 6) != 0;
+      this.qfi = (byte) (pppAndRqiAndQfi & 0x3F);
+
+      if (ppp) {
+        byte ppiAndSpare = ByteArrays.getByte(rawData, PPI_AND_SPARE_OFFSET + offset);
+        this.ppi = (byte) ((ppiAndSpare & 0xF0) >>> 4);
+        int paddingAndNextExtensionHeaderType =
+            ByteArrays.getInt(rawData, PADDING_AND_NEXT_EXTENSION_HEADER_TYPE_OFFSET + offset);
+        this.padding = (paddingAndNextExtensionHeaderType & 0xFFFFFF00) >>> 8;
+        this.nextExtensionHeaderType =
+            GtpV1ExtensionHeaderType.getInstance(
+                (byte) (paddingAndNextExtensionHeaderType & 0x000000FF));
+      } else {
+        this.nextExtensionHeaderType =
+            GtpV1ExtensionHeaderType.getInstance(
+                ByteArrays.getByte(rawData, NEXT_EXTENSION_HEADER_TYPE_OFFSET + offset));
+      }
+    } else if (pduType == 1) {
+      byte spareAndQfi = ByteArrays.getByte(rawData, SPARE_AND_QFI_OFFSET + offset);
+      this.qfi = (byte) (spareAndQfi & 0x3F);
+      this.nextExtensionHeaderType =
+          GtpV1ExtensionHeaderType.getInstance(
+              ByteArrays.getByte(rawData, NEXT_EXTENSION_HEADER_TYPE_OFFSET + offset));
+    }
+  }
+
+  private GtpV1PduSessionContainerExtensionHeader(Builder builder) {
+    this.pduType = builder.pduType;
+    this.ppp = builder.ppp;
+    this.rqi = builder.rqi;
+    this.qfi = builder.qfi;
+    this.ppi = builder.ppi;
+    this.padding = builder.padding;
+    this.nextExtensionHeaderType = builder.nextExtensionHeaderType;
+
+    if (builder.correctLengthAtBuild) {
+      this.length = 1;
+      if (pduType == 0 && ppp) {
+        this.length = 2;

```

Comment:
We'd better count padding length and calculate this length.

Reply:
It's already been taken account, in case “ppp” is true, the padding must be 3 bytes long so the “length” field should be 2.